Tutorial - Nuxt surveys (#7525)
PostHog/posthog.com: contents/tutorials/nuxt-surveys.md

---
title: How to set up surveys in Nuxt

> date: 2024-01-15

Author: ```suggestion
date: 2024-01-18
```

import EventsDark from '../images/tutorials/nuxt-surveys/events-dark.png'
import ImgSurveyResultsLight from '../images/tutorials/nuxt-surveys/survey-results-light.png'
import ImgSurveyResultsDark from '../images/tutorials/nuxt-surveys/survey-results-dark.png'
import ImgSurveyTemplatesLight from '../images/tutorials/nuxt-surveys/survey-templates-light.png'
import ImgSurveyTemplatesDark from '../images/tutorials/nuxt-surveys/survey-templates-dark.png'

[Surveys](/docs/surveys) are a great way to get feedback from your users. In this guide, we show you how to add a survey to your Nuxt.js app.

We'll create a basic Nuxt app, add PostHog, create a survey, and then show you how to display the survey in the app and get responses.

## 1. Create a Nuxt app

For this tutorial, we create a basic `Nuxt 3` app. First, ensure [Node.js is installed](https://nodejs.dev/en/learn/how-to-install-nodejs/) (version 14.6.0 or newer). Then run the following command:

```bash
npx nuxi@latest init <project-name>
```

Name it whatever you like (we call ours `nuxt-surveys`), select `npm` as the package manager, and use the defaults for the remaining options.

> Replace the code in `app.vue` with a simple heading and button:

Reviewer: ```suggestion
Replace the code in `app.vue` with a simple heading:
```

```vue file=app.vue
<template>
  <main>
    <h1>Nuxt.js 3 Surveys</h1>
  </main>
</template>
```

Run `npm run dev` to start your app.

![Basic app](../images/tutorials/nuxt-surveys/basic-app.png)

## 2. Add PostHog

> This tutorial shows how to integrate PostHog with `Nuxt 3`. If you're using `Nuxt 2`, see [our Nuxt docs](/docs/libraries/nuxt-js) for how to integrate PostHog.

We use PostHog to create and control our survey as well as monitor results. If you don't have a PostHog instance, you can [sign up for free here](https://us.posthog.com/signup). 

> First install `posthog-js`:

Reviewer: ```suggestion
To set it up with your Nuxt app, first, install `posthog-js`:
```

```ts file=nuxt.config.ts
export default defineNuxtConfig({
  // ...rest of your config
  
  runtimeConfig: {
    public: {
      posthogPublicKey: '<ph_project_api_key>',
      posthogHost: '<ph_instance_address>'
    }
  }
})
```

Create a new [plugin](https://nuxt.com/docs/guide/directory-structure/plugins) by first creating a new folder called `plugins` and then a new file `posthog.client.js`:

```bash
mkdir plugins
cd plugins 
touch posthog.client.js
```

Add the following code to your `posthog.client.js` file:

```js file=plugins/posthog.client.js
import { defineNuxtPlugin } from '#app'
import posthog from 'posthog-js'

export default defineNuxtPlugin(nuxtApp => {
  const runtimeConfig = useRuntimeConfig();
  const posthogClient = posthog.init(runtimeConfig.public.posthogPublicKey, {
    api_host: runtimeConfig.public.posthogHost,
  })

  // Make sure that pageviews are captured with each route change
  const router = useRouter();
  router.afterEach((to) => {
    posthog.capture('$pageview', {
      current_url: to.fullPath
    });

>   });

Reviewer: Do we need this for this tutorial? Even if it acts as a SPA, should capture at least an initial pageview on first load to prove it is working.

If you don't remove it, should at `capture_pageview: false` to initialization config.

Author: Yeah fair enough. Will remove

>   // Make sure that pageviews are captured with each route change
>   const router = useRouter();
>   router.afterEach((to) => {
>     posthog.capture('$pageview', {
>       current_url: to.fullPath
>     });
>   });

Author: ```suggestion
```

 
To create a survey with a prebuilt UI, go to the [surveys tab](https://us.posthog.com/surveys) in PostHog and click "New survey". 

<ProductScreenshot
  imageLight={ImgSurveyTemplatesLight} 
  imageDark={ImgSurveyTemplatesDark} 
  alt="PostHog survey templates" 
  classes="rounded"
/>

Select any template, or you can create your own by clicking "Create blank survey". Then, configure your survey with the following details:

1. Ensure `Presentation` is set to **Popover**.
2. Set the targeting to `All users`.
3. Use the default values for everything else.

Then, click "Save as draft" and then "Launch". Your survey is now live and you should see it in your app. After submitting responses, you can [view results in PostHog](#4-view-results).

![Popover survey in app](../images/tutorials/nuxt-surveys/popover-survey.png)

### Option 2: Implement your own survey UI

If you prefer to have complete control of your survey UI and logic, you can still use PostHog to keep track of and analyze your results.

> First, create a survey in PostHog like in option 1 above (for this tutorial, we use a Net promoter score survey template). The only difference is you must set `Presentation` to **API**.

Reviewer: ```suggestion
First, create a survey in PostHog like in option 1 above (for this tutorial, we use a net promoter score survey template). The only difference is you must set `Presentation` to **API**.
```

Then, there are four parts to adding code for our custom survey:

1. Create the survey UI.
2. Fetch the survey from PostHog.
3. Add the logic for displaying and hiding it.
4. Capture interactions from it.

#### 1. Create the survey UI

> We've created a sample survey UI for this tutorial. To use it, create a new file in `components` folder called `CustomSurvey.vue` and paste the following code:

Reviewer: Where does the components folder go? Base directory?

> We've created a sample survey UI for this tutorial. To use it, create a new file in `components` folder called `CustomSurvey.vue` and paste the following code:

Author: ```suggestion
We've created a sample survey UI for this tutorial. To use it, create a new folder `components` in your base directory and then create a new file `CustomSurvey.vue` in it. Paste the following code in this file:
```

});

const emit = defineEmits(['onDismiss', 'onSubmit']);

const selectedValue = ref(null);

function handleSelect(value) {
  selectedValue.value = value;
}

function emitDismiss() {
  emit('onDismiss');
}

function emitSubmit() {
  emit('onSubmit', selectedValue.value);
}
</script>

<style scoped>
.survey {
  position: fixed;
  bottom: 20px;
  right: 20px;
  width: 400px;
  padding: 20px;
  background-color: #ffffff;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.2);
  border-radius: 5px;
  z-index: 1000;
}

.button {
  margin: 5px;
  padding: 5px;
}
</style>
```

Then, integrate the component into `app.vue`:

```vue file=app.vue
<template>
  <main>
    <div class="App">
      <h1>Nuxt.js 3 Surveys</h1>
      <CustomSurvey
        v-if="showSurvey"
        :title="surveyTitle"
        @onDismiss="handleDismiss"
        @onSubmit="handleSubmit"
      />
    </div>
  </main>
</template>

<script setup>

import { ref } from 'vue';

const showSurvey = ref(true);
const surveyTitle = ref('Placeholder title')

const handleDismiss = () => {
  showSurvey.value = false;
};

const handleSubmit = (value) => {
  console.log("Submitted value:", value);
  showSurvey.value = false;
};

> 

Reviewer: ```suggestion
```

const surveyID = ref('');

onMounted(async () => {
  await fetchActiveSurveys();
});

const fetchActiveSurveys = () => {
  const { $posthog } = useNuxtApp();
    $posthog().getActiveMatchingSurveys((surveys) => {
    if (surveys.length > 0) {
      const survey = surveys[0];
      surveyID.value = survey.id;
      surveyTitle.value = survey.questions[0].question;
    }
  });
};

// rest of your code

</script>
```

#### 3. Add the logic for displaying and hiding it.

We want to make sure we don't show the survey again to users who have either submitted or dismissed it. We use [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) to store this data and use it to check whether to show the survey or not.

```vue file=app.vue
<template>
  <!-- ... rest of your template ... -->
</template>

<script setup>

import { ref, onMounted } from 'vue';
import { useNuxtApp } from '#app';

const showSurvey = ref(false); // updated to false

// rest of your code

const checkSurveyInteraction = () => {
  const hasInteractedWithSurvey = localStorage.getItem(`hasInteractedWithSurvey_${surveyID.value}`);
  showSurvey.value = !hasInteractedWithSurvey;
};

> onMounted(async () => {

Reviewer: Should mounted come before any methods?

Author: Its not necessary

There are 3 events to capture:

1. `"survey shown"`
2. `"survey dismissed"`
3. `"survey sent"` (for responses)

You can capture these events using `this.$posthog.capture()`:

```vue file=app.vue
<template>
  <!-- ... rest of your template ... -->
</template>

<script setup>

// rest of your code

const fetchActiveSurveys = () => {
  const { $posthog } = useNuxtApp();
    $posthog().getActiveMatchingSurveys((surveys) => {
    if (surveys.length > 0) {
      const survey = surveys[0];
      surveyID.value = survey.id;
      surveyTitle.value = survey.questions[0].question;
      checkSurveyInteraction();
      if (showSurvey) {
        $posthog().capture("survey seen", {
          $survey_id: surveyID // required
        })
      }
    }
  });
};

const handleDismiss = () => {
  showSurvey.value = false;
  localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');

  const { $posthog } = useNuxtApp();
  $posthog().capture("survey dismissed", {
    $survey_id: surveyID // required
  });
};

const handleSubmit = (value) => {
  showSurvey.value = false;
  localStorage.setItem(`hasInteractedWithSurvey_${surveyID}`, 'true');

  const { $posthog } = useNuxtApp();

>   $posthog().capture("survey sent", {

Reviewer: Just checked this event and it is showing this error:
![CleanShot 2024-01-16 at 20 16 37@2x](https://github.com/PostHog/posthog.com/assets/34755028/1a1516a8-94d4-4de7-8e5e-75eec5db238a)

Also not showing as a response in results, might be something to check to make sure it works properly.

Author: Interestingly this is a `number` on my PostHog instance. Maybe it was string and then later changed to number? @neilkakkar any idea what the correct type is supposed to be? 

<img width="750" alt="Screenshot 2024-01-17 at 8 10 10 PM" src="https://github.com/PostHog/posthog.com/assets/7090054/29bd0fd6-aa9e-4122-a8c7-1a9d70969e34">

Author: From the [slack conversation](https://posthog.slack.com/archives/C034XD440RK/p1705568310771809), this should be updated to number. I'll do this in a different PR, as there are a few other tutorials this will need to be updated for too (and also the docs)
